### PR TITLE
kvrpc: check raftcmd term.

### PR DESCRIPTION
**Pull request by disksing**
Add term in `kvrpc.Context`, add `StaleTerm` error.

For https://github.com/pingcap/tikv/issues/1317

### CODE REVIEW

File: proto/errorpb.proto [hunk 1]
```diff
@@ -31,11 +31,15 @@ message StaleEpoch {
 message ServerIsBusy {
 }
 
+message StaleTerm {
+}
```
Comment (siddontang) [inline proto/errorpb.proto]: need to return the origin and current term ?

Comment (disksing) [inline proto/errorpb.proto]: Not necessary. Only need to know the term is out of date.

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
I think we can use NotLeader error here.

If the peer A is leader at term n, and then it becomes leader again at term n + 1, it can still handle the request properly, no need to return error; If it loses leadership at term n + 1, and becomes leader again after term n + 1, then it was not leader in the past, so it's fine to return NotLeader error when receive a request with old term. And it can correct the leadership in TiDB client's cache too.

**Comment by disksing** [comment]
It would be strange if tikv server returns `NotLeader` but it actually is the leader.
And the cache correction is trivial, the cache will always be updated when we send a request to wrong node.

**Comment by BusyJay** [comment]
`NotLeader` error is not meant to say that the peer is not leader. TiKV can report `NotLeader` error even when current peer is a leader (ref: https://github.com/pingcap/tikv/blob/master/src/raftstore/store/peer.rs#L678). Maybe we can change the error name from `NotLeader` to `LeadershipChanged`.

**Comment by disksing** [comment]
I think it's better to let `NotLeader` always mean not leader, while change `StaleTerm` and others to `StaleCommand`.

**Comment by BusyJay** [comment]
Sounds good to me. But I think `LeadershipChanged` is more clear and precise.

**Comment by disksing** [comment]
Can't agree with `LeadershipChanged`...
`NotLeader` looks simple and straightforward if it always means 'not leader'.

**Comment by BusyJay** [comment]
I mean changing `StaleTerm` and others to `LeadershipChanged`.

**Comment by disksing** [comment]
Still not agree. You may search for usages of `notify_not_leader`, direct reason of calling it is about command stale (caused by leadership change indirectly).

**Comment by BusyJay** [comment]
All the usages other than actual `NotLeader` are due to leadership change. What is a stale command? How a command becomes stale? These are unanswered questions.

**Comment by disksing** [comment]
Error name does not need to answer the deep reason that cause it. For instance, `StaleEpoch` does not describe how the epoch becomes stale.

**Comment by BusyJay** [comment]
I don't think so. For example, `NotLeader` is the reason that a command can't be handled. `LeadershipChanged` is the same as `NotLeader`. `StaleEpoch` indicates the command's epoch doesn't match the epoch the peer currently holds, hence it can't handle the command. But what does `StaleCommand` mean? What does stale mean? How does a command become stale? This is very confusing and obscure.

**Comment by disksing** [comment]
Let's put the 2 choices into old code:

```
            while let Some(cmd) = self.pending_cmds.pop_normal(term - 1) {
                // apprently, all the callbacks whose term is less than entry's term are stale.
                self.notify_leadership_changed(cmd);
            }
```

```
            while let Some(cmd) = self.pending_cmds.pop_normal(term - 1) {
                // apprently, all the callbacks whose term is less than entry's term are stale.
                self.notify_stale_command(cmd);
            }
```

I think your suggestion makes it more confuse.

**Comment by BusyJay** [comment]
How? The command with less term are belongs to old leader, hence `apparently`. So `notify_leadership_changed` here looks good to me and is straightforward.

**Comment by siddontang** [comment]
I think StaleTerm is clear. 

/cc @ngaut

**Comment by ngaut** [comment]
stale_command LGTM

**Comment by disksing** [comment]
PTAL

**Comment by siddontang** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**